src/query/storage/promremote/storage.go

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2021  Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package promremote
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/uber-go/tally"
+
+	"github.com/m3db/m3/src/query/storage"
+	xerrors "github.com/m3db/m3/src/x/errors"
+	"github.com/m3db/m3/src/x/instrument"
+	xhttp "github.com/m3db/m3/src/x/net/http"
+)
+
+const metricsScope = "prom_remote_storage"
+
+// NewStorage returns new Prometheus remote write compatible storage
+func NewStorage(opts Options) (storage.Storage, error) {
+	client := xhttp.NewHTTPClient(opts.httpOptions)
+	scope := opts.scope.SubScope(metricsScope)
+	s := &promStorage{opts: opts, client: client, endpointMetrics: initEndpointMetrics(opts.endpoints, scope)}
+	return s, nil
+}
+
+type promStorage struct {
+	opts            Options
+	client          *http.Client
+	endpointMetrics map[string]*instrument.MethodMetrics
+}
+
+func (p *promStorage) Write(ctx context.Context, query *storage.WriteQuery) error {
+	encoded, err := encodeWriteQuery(query)
+	if err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	multiErr := xerrors.NewMultiError()
+	var errLock sync.Mutex
+	for _, endpoint := range p.opts.endpoints {
```

Comment:
We need to clarify the requirements to make sure we don't have to return an error in case no matching endpoint was found. Maybe there should be an option to specify some default "catch everything else" endpoint.

Reply:
Good idea. Will followup with a PR.

Reply:
+1 let's perhaps return an error here for the time being?

I don't think the default catch else namespace is actually all that compelling, but returning an error would be ideal (this is what we do for m3 storage).